File: Peetcode/LongestCommonSequence.py
def LongestCommonSequence(str1,str2):
    m,n = len(str1),len(str2)
    record = [[0 for i in range(n+1)] for j in range(m+1)]
    for i in range(1,m+1):
        for j in range(1,n+1):
            if str1[i-1]==str2[j-1]:
                record[i][j] = record[i-1][j-1] + 1
            elif record[i-1][j]>=record[i][j-1]:
                record[i][j] = record[i-1][j]
            else:
                record[i][j] = record[i][j-1]
            # 这里不要使用以下语句，因为打印序列的时候无法处理相等情况下的方向
            # record[i][j] = max(record[i-1][j],record[i][j-1])
    return record

def printLCS(x,y,c):
    # 这里的 c 就是上面的 record 矩阵
    s=[]
    i=len(x)
    j=len(y)
    while i>0 and j>0:
        if x[i-1]==y[j-1]:
            s.append(x[i-1])
            i=i-1
            j=j-1
        elif c[i][j-1]>=c[i-1][j]:
            j=j-1
        else:
            i=i-1
    print("Longest Common Subsequence is:" ,list(reversed(s)))

if __name__=='__main__':
    # print(LongestCommonSequence("abcughigh","bdcuihghvh"))
    print(printLCS("abcughigh","bdcuihghvh",LongestCommonSequence("abcughigh", "bdcuihghvh")))

File: Ceetcode/动态规划/LongestCommonSequence.cpp
/* 最长公共子序列：经典的动态规划题目，时常复习，解法经典 */
/* 下面是测试通过的案例 */

int longestCommonSequence(string &A, string &B) {
	int m = A.size(), n = B.size();
	if (m < n)
		return longestCommonSequence(B, A); // 保证有最小的空间复杂度

	vector<int> dp(n + 1, 0);
	vector<int> dp2(n + 1, 0);
	for (int i = 1; i <= m; i++) {
		for (int j = 1; j <= n; j++) {
			dp[j] = (A[i - 1] == B[j - 1]) ? dp2[j - 1] + 1 : max(dp[j - 1], dp2[j]);
		}
		swap(dp, dp2);
	}
	return dp2[n];
}

